test/module/irohad/ametsuchi/flat_file_block_storage_test.cpp

```diff
@@ -0,0 +1,176 @@
+                               std::string> {
+        auto return_block = std::make_unique<MockBlock>();
+        EXPECT_CALL(*return_block, height()).WillOnce(Return(height_));
+        return iroha::expected::makeValue<
+            std::unique_ptr<shared_model::interface::Block>>(
+            std::move(return_block));
+      }));
+  std::shared_ptr<const shared_model::interface::Block> block_var =
+      block_storage->fetch(height_).get();
+
+  ASSERT_EQ(block_->height(), block_var->height());
+}
+
+/**
+ * @given initialized block storage without blocks
+ * @when block with height_ is fetched
+ * @then nothing is returned
+ */
+TEST_F(FlatFileBlockStorageTest, FetchNonexistent) {
+  auto block_storage =
+      FlatFileBlockStorageFactory(path_provider_, converter_, log_manager_)
+          .create();
+  auto block_var = block_storage->fetch(height_);
+  ASSERT_FALSE(block_var);
+}
+
+/**
+ * @given initialized block storage, single block with height_ inserted
+ * @when size is fetched
+ * @then 1 is returned
+ */
+TEST_F(FlatFileBlockStorageTest, Size) {
+  auto block_storage =
+      FlatFileBlockStorageFactory(path_provider_, converter_, log_manager_)
+          .create();
+  ASSERT_TRUE(block_storage->insert(block_));
+
+  ASSERT_EQ(1, block_storage->size());
+}
+
+/**
+ * @given initialized block storage, single block with height_ inserted
+ * @when storage is cleared with clear
+ * @then no blocks are left in storage
+ */
+TEST_F(FlatFileBlockStorageTest, Clear) {
+  auto block_storage =
+      FlatFileBlockStorageFactory(path_provider_, converter_, log_manager_)
+          .create();
+  ASSERT_TRUE(block_storage->insert(block_));
+
+  block_storage->clear();
+
+  auto block_var = block_storage->fetch(height_);
+
+  ASSERT_FALSE(block_var);
+}
+
+/**
+ * @given initialized block storage, single block with height_ inserted
+ * @when forEach is called
+ * @then block with height_ is visited, lambda is invoked once
+ */
+TEST_F(FlatFileBlockStorageTest, ForEach) {
+  auto block_storage =
+      FlatFileBlockStorageFactory(path_provider_, converter_, log_manager_)
+          .create();
+  ASSERT_TRUE(block_storage->insert(block_));
+
+  EXPECT_CALL(*converter_, deserialize(_))
+      .WillOnce(Invoke([&](const shared_model::interface::types::JsonType &)
+                           -> iroha::expected::Result<
+                               std::unique_ptr<shared_model::interface::Block>,
+                               std::string> {
+        auto return_block = std::make_unique<MockBlock>();
+        EXPECT_CALL(*return_block, height()).WillOnce(Return(height_));
+        return iroha::expected::makeValue<
+            std::unique_ptr<shared_model::interface::Block>>(
+            std::move(return_block));
+      }));
+
+  size_t count = 0;
+
+  block_storage->forEach([this, &count](const auto &block) {
+    ++count;
+    ASSERT_EQ(height_, block->height());
```

Comment:
Cannot we check the block pointer equivalence here instead of height? Something like:
```suggestion
    ASSERT_EQ(expected_block, block);
```
where `expected_block` was created higher and returned by mock `deserialize_` call?

Reply:
Technically we can. In practice these pointers are never expected to be the same so such equivalence would be confusing in my opinion.

Reply:
why do we not require the exact block returned from deserializer to appear in visitor argument? as I can see, deserializer works like a block creator there.

Reply:
Well, that's a good point. The problem of comparing whole blocks is just that we will need to initialise all fields which makes comparing mock blocks basically a comparing of real blocks. Also that's pretty painful to do.

Reply:
I meant pointers comparison

Reply:
The problem here is `deserialize` call which returns `unique_ptr` and not a `shared_ptr`. I have no idea how to write mock expectation for this case (and tbh I'm not sure if it's even possible).

Reply:
let's just compare raw pointer values, like it is done in other places

Reply:
Looks like it's a bit problematic: we still have to create `unique_ptr` inside of `deserialize` except and if we construct it from a raw pointer it will be double destroyed occasionally.

Reply:
I do not get the problem. You already have two pointers: the one created here
https://github.com/hyperledger/iroha/blob/9303196deea08b2e1ff0e7093bebd5c23e90ed86/test/module/irohad/ametsuchi/flat_file_block_storage_test.cpp#L161
and the one got here
https://github.com/hyperledger/iroha/blob/9303196deea08b2e1ff0e7093bebd5c23e90ed86/test/module/irohad/ametsuchi/flat_file_block_storage_test.cpp#L170
which should store equal raw pointers.
please correct me if I am missing something